stream/src/public_key/handshake.rs

```diff
@@ -124,6 +124,15 @@ impl<C: Scheme> Signed<C> {
             return Err(Error::HandshakeNotForUs);
         }
 
+        // Verify that the handshake is not signed by us
+        //
+        // This could indicate a self-connection attempt, which is not allowed.
+        // It could also indicate a replay attack or a malformed message.
+        // Either way, fail early to avoid any potential issues.
+        if crypto.public_key() == self.signer {
```

Comment:
Unlike in `dialer`, definitely not safe to `assert` here (as input chosen by peer). Just calling out for symmetry (no change suggested).